Unit tests for some more straightforward steps

petr-muller opened the pull request:
Before trying to unit test some of the larger, complicated steps that would actually need to mock some cluster state, I'm trying to cover some of the simpler ones. I have tried to extract some testing routines to a separate `testing.go` file and I'm trying to use that for testing all the steps. I needed to grok what `DeferredParams` in `template.go` does, so I wrote unit tests for it, discovering some suspicious parts in the process:

- Map() method does not perform the `os.LookupEnv(name)` like the other read-ish methods do, so `Map()` can return something not consistent with what `Get()` would return.
- `DeferredParams.links` is a slice (`[]api.StepLink)`, but it does not ever contain more than a single element. The only method that writes it is `Add()` and that only sets it to be a single-item slice. So either this is wrong or it could be just a straight `api.StepLink`, not a slice.

I'm mostly submitting this to get opinions on the `testing.go` stuff and how the actual tests look like when using that code.

petr-muller commented:
/cc @stevekuznetsov @droslean

stevekuznetsov commented on pkg/steps/template_test.go:
```diff
@@ -0,0 +1,173 @@
+				"K2": {someStepLink("THREE")},
+			},
+		},
+		// only compare count of returned items because AllLinks can
```
Can sort for deterministic ordering

petr-muller replied:
Not that simply, seems like I would need to implement some interface to make those classes sortable and I was lazy. I'll consider it (and maybe I'll even get rid of misusing the "real" StepLink structs and implement a simple one for testing).

petr-muller replied:
IMO not worth the effort, `StepLink` is an interface that can do only `Same` and `Matches`, there's nothing to sort them by easily :/

stevekuznetsov commented on pkg/steps/template_test.go:
```diff
@@ -0,0 +1,173 @@
+	}
+}
+
+func TestDeferredParametersMap(t *testing.T) {
```
Not being super familiar with the underlying code, I'm having trouble understanding the test. If we had more than one case it would be useful here, also a description/name that provided setup --> end result causality.

Is this test case ensuring that a func doesn't override a previous value? Does it enforce that a previous value without a func is not exposed? Are there other cases? Can we break them up into separate cases?

petr-muller replied:
I'll try to look into this to make it more readable, at least with descriptions and/or names.

petr-muller replied:
I've broken the testcase to smaller, more targeted ones. You were right I was probably trying to do too much in a single one.

stevekuznetsov commented on pkg/steps/template_test.go:
```diff
@@ -0,0 +1,173 @@
+		links: []api.StepLink{someStepLink("new-link")},
+	}}
+	for _, tc := range testCases {
+		if tc.name != "DO-NOT-ADD" {
```
This smells of a bad test array setup?

petr-muller replied:
Indeed :)

petr-muller replied:
Replaced with explicit `callAdd` / `callSet` booleans.

stevekuznetsov commented on pkg/steps/template_test.go:
```diff
@@ -0,0 +1,173 @@
+		getValue: "oldValue",
+		getError: nil,
+	}, {
+		purpose: "Existing key is not set if lazy evaluation func is set",
```
Ok, these names rock :100:

petr-muller commented:
Thanks for the feedback, I'll try to incorporate it. I was working on testing the `test.go` step in the meantime, which finally does some testing against fakes, and I did some changes, so I need to reconcile all this work into some sensible changes.

petr-muller commented:
I've addressed the feedback and also incorporated some followup work, mainly the fact that I've separated the original `examineStep` which tested all `Step` methods into two smaller ones. The `examineStep` checks the "static" methods (not related to step execution) and `executeStep` that checks the `Done()` and `Run()` methods. The rationale for that is that is that especially `Run()` is likely to be checked in different scenarios for some steps (different cluster behavior, different inputs etc) and it is useless to check everything all the time.

Ready for re-review.

stevekuznetsov commented on pkg/steps/template_test.go:
```diff
@@ -0,0 +1,220 @@
+	"testing"
+
+	"github.com/openshift/ci-operator/pkg/api"
+	"k8s.io/apimachinery/pkg/util/diff"
```
nit: group imports, first stdlib, then third party, then ci-operator

petr-muller replied:
Addressed.

stevekuznetsov commented:
/lgtm
